sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/read/UnboundedSolaceReader.java

```diff
@@ -54,11 +56,23 @@ class UnboundedSolaceReader<T> extends UnboundedReader<T> {
   AtomicBoolean active = new AtomicBoolean(true);
 
   /**
-   * Queue to place advanced messages before {@link #getCheckpointMark()} be called non-concurrent
-   * queue, should only be accessed by the reader thread A given {@link UnboundedReader} object will
-   * only be accessed by a single thread at once.
+   * List of successfully ACKed message (surrogate) ids which need to be pruned from the above.
+   * CAUTION: Accessed by both reader and checkpointing threads.
    */
-  private final java.util.Queue<BytesXMLMessage> elementsToCheckpoint = new ArrayDeque<>();
+  private final Queue<Long> ackedMessageIds;
+
+  /**
+   * Map to place advanced messages before {@link #getCheckpointMark()} is called. This is a
+   * non-concurrent object, should only be accessed by the reader thread.
+   */
+  private final Map<Long, BytesXMLMessage> safeToAckMessages;
+
+  /**
+   * Surrogate id used as a key in Collections storing messages that are waiting to be acknowledged
+   * ({@link UnboundedSolaceReader#safeToAckMessages}) and already acknowledged ({@link
+   * UnboundedSolaceReader#ackedMessageIds}).
+   */
+  private Long surrogateId = 0L;
```

Comment:
It seems like you could use the queue directly instead of this map.
The reader can offer messages to the queue, while your checkpoint mark can poll from it and subsequently call `ackMessage()`. The elements observed by this reader should be finalized by checkpoint marks of this reader. Alternatively, a deserialized checkpoint mark may be provided the current reader's queue to perform clean up. The reference to this reader's queue may need to be marked as volatile, but I'd have to double check when the checkpoint mark is handed off to a different thread.